src/common/Validation.ml

```diff
@@ -110,58 +104,71 @@ end
   type t
-end) : S with type error := X.t = struct
-  type errors = NonEmpty of X.t * X.t list
+end) (Error : sig
+  type t
+end) : S with module Warning := Warning and module Error := Error = struct
+  type 'a t = ('a * Warning.t list, Error.t list * Warning.t list) result
+
+  module Basic = struct
+    type nonrec 'a t = 'a t
+
+    let map_ x ~f =
+      match x with
+      | Result.Ok (x, ws) -> Ok (f x, ws)
+      | Error (x, ws) -> Error (x, ws)
+
+    let map = `Custom map_
 
-  let append xs ys =
-    match (xs, ys) with
-    | NonEmpty (x, []), NonEmpty (y, []) -> NonEmpty (x, [y])
-    | NonEmpty (x, xs), NonEmpty (y, ys) -> NonEmpty (x, xs @ (y :: ys))
+    let apply f x =
+      match (f, x) with
+      | Result.Ok (f, ws), Result.Ok (x, vs) -> Result.Ok (f x, ws @ vs)
+      | Ok (_, ws), Error (e, vs) -> Error (e, ws @ vs)
+      | Error (e, ws), Ok (_, vs) -> Error (e, ws @ vs)
+      | Error (es, ws), Error (fs, vs) -> Error (es @ fs, ws @ vs)
 
-  type 'a t = ('a, errors) result
+    let bind x ~f =
+      match x with
+      | Result.Ok (x, ws) -> (
+        match f x with
+        | Result.Ok (x, vs) -> Result.Ok (x, ws @ vs)
+        | Error (e, vs) -> Error (e, ws @ vs) )
+      | Error (e, ws) -> Error (e, ws)
 
-  let map x ~f = match x with Ok x -> Ok (f x) | Error x -> Error x
-  let pure x = Ok x
+    let return x = Result.Ok (x, [])
+  end
 
-  let apply x ~f =
-    match (f, x) with
-    | Ok f, Ok x -> Ok (f x)
-    | Error e, Ok _ -> Error e
-    | Ok _, Error e -> Error e
-    | Error e1, Error e2 -> Error (append e1 e2)
+  include Applicative.Make (Basic)
+  include Monad.Make (Basic)
 
-  let apply_const a b = apply b ~f:(map a ~f:(fun _ x -> x))
-  let bind x ~f = match x with Ok x -> f x | Error e -> Error e
-  let liftA2 f x y = apply y ~f:(apply x ~f:(pure f))
-  let liftA3 f x y z = apply z ~f:(apply y ~f:(apply x ~f:(pure f)))
-  let consA next rest = liftA2 List.cons next rest
-  let sequence ts = List.fold_right ~init:(pure []) ~f:consA ts
+  let apply_const a b = apply (map a ~f:(fun _ x -> x)) b
+  let liftA2 f x y = apply (apply (return f) x) y
+  let liftA3 f x y z = apply (apply (apply (return f) x) y) z
+  let ok x = return x
 
-  module Validation_infix = struct let ( >>= ) x f = bind x ~f end
-  include Validation_infix
+  let error ?warn err : 'a t =
+    Result.Error ([err], Option.value_map ~default:[] ~f:(fun x -> [x]) warn)
 
-  let ok x = pure x
-  let error x = Error (NonEmpty (x, []))
-  let is_error = function Error _ -> true | _ -> false
-  let is_success = function Ok _ -> true | _ -> false
+  let warn ~warn x : 'a t = Result.Ok (x, [warn])
+  let is_error = function Result.Error _ -> true | _ -> false
+  let is_warning = function Result.Ok (_, _ :: _) -> true | _ -> false
+  let is_success = function Result.Ok (_, []) -> true | _ -> false
 
   let get_errors_opt = function
-    | Error (NonEmpty (x, xs)) -> Some (x :: xs)
+    | Result.Error (es, ws) -> Some (es, ws)
```

Comment:
What's the use of these custom accessor functions that stac the `option` type on here? A List can be empty and has its own library of functions for accessing various parts of it.

Reply:
I don't think this is needed at all; we just need `to_result` I think.